dispatcher/index.js

```diff
@@ -0,0 +1,155 @@
+  }
+
+  function removeHandler(handler, keypaths) {
+    for (var i = 0, length = handlers.length; i < length; i++) {
+      if (handlers[i][0] === handler && handlers[i][1] === keypaths) {
+        handlers = handlers.slice(0, i).concat(handlers.slice(i + 1));
+        return;
+      }
+    }
+  }
+
+  function dispatchActionOnHandler(action, data, handler, keypaths, actionQueue, handlePromise) {
+    var newData;
+
+    if (!keypaths) {
+      newData = handler(action, data);
+    } else {
+      var firstKeyPath = keypaths[0];
+      if (keypaths.length === 1) {
+        newData = handler(action, get(data, firstKeyPath));
+      } else {
+        var subTrees = keypaths.map(resolveKeypathOnThis, data);
+        newData = handler.apply(null, [action].concat(subTrees));
+      }
+    }
+
+    if (isIterator(newData)) {
+      var iterator = newData;
+      var next = iterator.next();
+      newData = next.value;
+      consumeIterator(iterator, actionQueue, handlePromise);
+    }
+
+    return (
+      newData === undefined ? data :
+      firstKeyPath ? set(data, firstKeyPath, newData) :
+      newData
+    );
+  }
+
+  function dispatch(actionQueue, pendingPromises, data, updatedData, callback) {
+    function handlePromise(promise, maybeUnfinishedIterator) {
+      pendingPromises.push(promise);
+      function continueDispatch() {
+        consumeIterator(maybeUnfinishedIterator, actionQueue, handlePromise);
+        removeFromArray(pendingPromises, promise);
+        runDispatch(updatedData ? updatedData() : data);
+      }
+
+      promise.then(function(action) {
+        actionQueue.push(action);
+        immediate(continueDispatch); // get out of promise error handling
+      });
+    }
+
+    function invokeHandler(spec, action, currentData) {
+      var handler = spec[0];
+      var keypaths = spec[1];
+      return dispatchActionOnHandler(action, currentData, handler, keypaths, actionQueue, handlePromise);
+    }
+
+    function invokeHandlers(action, currentData) {
+      return handlers.reduce(function(mutatedData, spec) {
+        return invokeHandler(spec, action, mutatedData);
+      }, currentData);
+    }
+
+    function runDispatch(currentData) {
+      var action;
+      while ((action = actionQueue.shift())) {
+        currentData = invokeHandlers(action, currentData);
+      }
+
+      var isDone = pendingPromises.length === 0;
+      if (callback && (isDone || !equals(data, currentData))) {
+        data = currentData;
+        callback(currentData, isDone);
+      }
+    }
+
+    runDispatch(data);
+  }
+
+  return {
+    register: function(handler) {
+      var keypaths = arguments.length < 2 ?
+        null : argumentsToKeypaths(arguments, 1);
+      handlers.push([handler, keypaths]);
+      return function remove() { removeHandler(handler, keypaths); };
+    },
+
+    dispatch: function(action, initialData, callback, updatedData) {
+      dispatch([action], [], initialData, updatedData, callback);
```

Comment:
The name `updatedData` is kind of confusing. `getUpdatedData` would be a more obvious name for it.